### ISSUE TITLE
Strange behavior of `ndimage.shift` and `ndimage.affine_transform`

### ISSUE THREAD

**Issue by MatthiasKohl** [scipy/scipy#9865]
My issue is about a strange behavior on the borders of `ndimage.shift`/`ndimage.affine_transform` when using bilinear interpolation and constant border mode.

It looks like these functions assume that if the input indexes lie in `[0, last_idx]` then the output coordinate system has to be `[0, last_idx]` and any point outside of this system in the output must be the constant value (here 0). I would have assumed that the output coordinate system is `[-0.5, last_idx + 0.5]` as this is what a pixel usually represents.

Can you clarify that this is the expected behavior and how to obtain the result as with convolution, but possibly with a faster implementation for larger-scale arrays ?

EDIT: https://github.com/scipy/scipy/issues/1323 may be related but it uses `mode=reflect`

### Reproducing code example:

```
from scipy.ndimage import shift, affine_transform
import numpy as np
e = np.eye(3)
e[(0, 2), (2, 0)] = 1
print(e)

print(shift(e, -.4, order=1, mode='constant'))
mat = np.array([[1, 0, .4], [0, 1, .4], [0, 0, 1]])
print(affine_transform(e, mat, order=1, mode='constant'))
```

### Output:

```
array([[1., 0., 1.],
       [0., 1., 0.],
       [1., 0., 1.]])

array([[0.52, 0.48, 0.  ],
       [0.48, 0.52, 0.  ],
       [0.  , 0.  , 0.  ]])

array([[0.52, 0.48, 0.  ],
       [0.48, 0.52, 0.  ],
       [0.  , 0.  , 0.  ]])
```

### Expected output:
```
from scipy.signal import convolve
k = np.array([[.16, .24, 0], [.24, .36, 0], [0, 0, 0]])
print(convolve(e, k, mode='same'))

array([[0.52, 0.48, 0.36],
       [0.48, 0.52, 0.24],
       [0.36, 0.24, 0.36]])
```

### Scipy/Numpy/Python version information:

1.2.0 1.15.4 sys.version_info(major=3, minor=6, micro=8, releaselevel='final', serial=0)

**Comment by MikhailRyazanov** [comment]
Perhaps an easier way to demonstrate that something is very wrong with the boundary conditions:
```python
import numpy as np
from scipy.ndimage import shift

a = np.ones(5)
b = shift(a, 1.5, order=1)
print(a, '--(1.5)->', b)

a[0] = 0
b = shift(a, 0.5, order=1)
print(a, '--(0.5)->', b)
```
```
[1. 1. 1. 1. 1.] --(1.5)-> [0. 0. 1. 1. 1.]
[0. 1. 1. 1. 1.] --(0.5)-> [0.  0.5 1.  1.  1. ]
```
The second result is correct, and the first, obviously, must be the same, but it is clearly not.

Moreover, with the default `order=3` the behavior is really strange:
```
[1. 1. 1. 1. 1.] --(1.5)-> [0. 0. 1. 1. 1.]
[0. 1. 1. 1. 1.] --(0.5)-> [0.         0.39955357 1.12723214 0.96651786 1.00669643]
```
The second result indeed looks like cubic interpolation, but the first does not look like “Points outside the boundaries of the **input** are filled [with zeros]” at all.

**Comment by grlee77** [comment]
closed by #12776, but note that mode `'grid-constant'` rather than `'constant'` has to be specified to get the expected behavior. The older constant mode behavior was retained for backwards compatibility.
